[BD-14] Implement basic features

arbrandes opened the pull request:
### Description:

Includes:

* Basic configuration of the MFE template (package name, service port, etc.)
* A Studio-like header
* A home page which lists both v1 and v2 libraries
* A library detail page
* A library creation form

### Manual Testing

To test this manually in a devstack, proceed as follows:

1. Check out [this devstack branch](https://github.com/edx/devstack/pull/556) and `make dev.clone`.

1.  After `make dev.clone` (or `make dev.checkout`, as needed), check out this PR's branch (`arbrandes/librariesv2-sagaless`) in `frontend-app-library-authoring`.

1. Launch the master devstack as usual.

1. Launch `frontend-app-library-authoring` via `make dev.up.frontend-app-library-authoring`, since it is not a default service.

1. Checkout out [blockstore](https://github.com/edx/blockstore/), `make dev.up` and `make dev.provision` it.  In `make blockstore-shell`, create an auth token for edx-platform and a new collection, making a note of its UUID:

    ```bash
    echo "from django.contrib.auth import get_user_model; from rest_framework.authtoken.models import Token; User = get_user_model(); edxapp_user, _ = User.objects.get_or_create(username='edxapp'); Token.objects.get_or_create(user=edxapp_user, key='edxapp-insecure-devstack-key')" | ./manage.py shell
    ./manage.py createcollection "Studio Collection"
    Created Collection: 36aefd6a-c39e-4ede-ab48-fe40ddecfb33 (Studio Collection)
    ```

1. Start blockstore with `make easyserver`.

1. Edit `frontend-app-library-authoring/.env.development` and add the UUID you obtained above to BLOCKSTORE_COLLECTION_UUID:

    ```bash
    BLOCKSTORE_COLLECTION_UUID=36aefd6a-c39e-4ede-ab48-fe40ddecfb33
    ```

1. Restart the `frontend-app-library-authoring` container:

    ```bash
    cd devstack
    docker-compose restart frontend-app-library-authoring
    ```

1. In the Studio shell, configure it with blockstore credentials and enable the ENABLE_LIBRARY_AUTHORING_MICROFRONTEND feature flag:

    ```bash
    make studio-shell
    echo "BLOCKSTORE_API_AUTH_TOKEN = 'edxapp-insecure-devstack-key'" >> /edx/app/edxapp/edx-platform/cms/envs/private.py
    echo "BLOCKSTORE_API_AUTH_TOKEN = 'edxapp-insecure-devstack-key'" >> /edx/app/edxapp/edx-platform/lms/envs/private.py
    vim /edx/etc/studio.yml
    ---
    FEATURES:
        ENABLE_LIBRARY_AUTHORING_MICROFRONTEND: true
    ```

1. Exit the shell and restart Studio:

    ```bash
    make studio-restart
    ```

1. In the browser, go to http://localhost:18010/admin/organizations/organization/add/, log in as `edx`, and create an organization named (and short-named)`edX`.

1. Finally, create a `studio.library_authoring_mfe` waffle flag via http://localhost:18010/admin/waffle/flag/, and enable it for everyone.

1. Now go to Studio Home (http://localhost:18010/home/) and click on the Libraries tab.

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_21-27-08](https://user-images.githubusercontent.com/759355/87212391-a5f00200-c2f4-11ea-80d9-91b0ff09b435.png)
    </details>

1. You should be taken to the MFE as a logged-in user and presented with an empty page.  (If, on the other hand, there are previously existing libraries, they'll be shown here).

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_21-27-59](https://user-images.githubusercontent.com/759355/87212405-b0120080-c2f4-11ea-8f6a-c7891f0b015f.png)
    </details>

1. Click on `New Library` in the upper right.  You should see the form appear.

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_21-35-13](https://user-images.githubusercontent.com/759355/87212525-5fe76e00-c2f5-11ea-8d3e-9bd29227f5df.png)
    </details>

1. Try to create a new complex library by filling out the form with a bogus organization and clicking `CREATE`.  Note that the button is not enabled until all required fields are filled out.  You'll get an error message as follows:

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_21-37-18](https://user-images.githubusercontent.com/759355/87212598-c1a7d800-c2f5-11ea-8af2-32fa4e32b78e.png)
    </details>

1. Change the organization name to "edX", and click `CREATE` again.  You should be taken to the placeholder library detail page.

    <details>
    <summary>Screenshot</summary>

     ![Screenshot_2020-07-10_22-03-26](https://user-images.githubusercontent.com/759355/87213156-519b5100-c2f9-11ea-9cfd-5cdaabe5f653.png)
    </details>

1. Click the Studio logo on the top left to go back to the library list page.  You should now see the new library there:

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_22-05-28](https://user-images.githubusercontent.com/759355/87213184-7f809580-c2f9-11ea-8d67-a5efe46101a4.png)
    </details>

1. Fill out the form to create a new _legacy_ library:

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_22-06-42](https://user-images.githubusercontent.com/759355/87213207-c3739a80-c2f9-11ea-96dd-6bd6ec144d6d.png)
    </details>

1. After clicking `CREATE`, you should be taken to the library's regular Studio page (as opposed to the MFE placeholder detail page):

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_22-08-29](https://user-images.githubusercontent.com/759355/87213237-f4ec6600-c2f9-11ea-8bdd-74cb5849caa5.png)
    </details>

1. Go back to https://localhost:3001.  Both libraries should be listed:

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_22-09-51](https://user-images.githubusercontent.com/759355/87213263-1fd6ba00-c2fa-11ea-8ff6-f5309b33e9c7.png)
    </details>

1. Finally, click on both libraries.  The complex library should take you to its detail page in the MFE, and the legacy one should take you to its page in Studio.

edx-webhook commented:
Thanks for the pull request, @arbrandes! I've created [BLENDED-537](https://openedx.atlassian.net/browse/BLENDED-537) to keep track of it in Jira.   More details are on the [BD-14](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1545011241/BD-14+Blockstore+Powered+Content+Libraries+Taxonomies) project page.

When this pull request is ready, tag your edX technical lead.

<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->

arbrandes commented:
@kdmccormick, I've recreated the open-craft repo by forking it from the edX one, so that pull requests would behave as usual.  This PR is a refactored version of the previous one you reviewed.

It:

* Removes redux-saga as suggested by @davidjoy (@bradenmacdonald, this also removes the generator functions you noted earlier)
* Ports the Studio index tabs to the main page, so that users have a natural way to get back to Studio
* Fixes the form submission bug where users would be taken to a previously successfully submitted form in case of a server-side validation error (reported by @mavidser)
* Uses a single status variable for the main components, as suggested by @mavidser
* Fixes a CSS bug where the FontAwesome plus icon would not be displayed on the create buttons

What is still does _not_ do, is:

* Introduce tests
* Make the user account dropdown in the header match the design language in Studio.  Last time around you mentioned you'd investigate the state of the Studio header in other MFEs, and get back to me.  I notice there's now a [studio-header](https://github.com/edx/frontend-app-course-authoring/tree/master/src/studio-header) in frontend-app-course-authoring, but I haven't tested it.

It has also not been:

* Copy-checked by @marcotuts 
* Accessibility audited

I intend to address the above omissions in future PRs, as you also suggested earlier.

kdmccormick commented:
Thanks @arbrandes! This is next in my queue of PRs to review.

> Last time around you mentioned you'd investigate the state of the Studio header in other MFEs, and get back to me. I notice there's now a studio-header in frontend-app-course-authoring, but I haven't tested it.

Yup -- I have nothing to show for this yet. There's a ticket in my team's next sprint to start addressing all this header business; I hope to have some update ready in time for our video sync.

> It has also not been:
> * Copy-checked by @marcotuts
> * Accessibility audited

I'll make tickets to track the work that needs to happen for these on the edX side. Although we'll need to do them eventually, let's continue to see them as non-blockers for development work.

> I intend to address the above omissions in future PRs, as you also suggested earlier.

Sounds good :+1:

kdmccormick commented on src/library/form/LibraryForm.jsx:
```diff
@@ -0,0 +1,251 @@
+
+import messages from './messages';
+
+class LibraryForm extends React.Component {
```
Would it be fair to be a bit more explicit and call this something like `LibraryCreateForm` or `CreateLibrayForm`?

arbrandes replied:
Done!

kdmccormick commented on src/library/list/LibraryItem.jsx:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Represents a library in the list of libraries.
+ */
+const LibraryItem = ({intl, library}) => (
```
How about `LibraryListItem`? Might make it even clearer what this class is without having to read the above comment.

arbrandes replied:
Done!

kdmccormick commented on src/library/header/LibraryHeader.jsx:
```diff
@@ -0,0 +1,99 @@
+  'LOGOUT_URL',
+], 'Library header');
+
+class LibraryHeader extends React.Component {
```
How about `LibraryAuthoringHeader`? Someone unfamiliar might guess that `LibraryHeader` refers to the header of an individual library.

arbrandes replied:
I called it StudioHeader, since that's what it is, or at least what it's intended to be. :shrug:

kdmccormick commented on src/library/index.js:
```diff
@@ -0,0 +1,4 @@
+export * from './detail';
+export * from './form';
+export * from './header';
+export * from './list';
```
Consider renaming these folders (including the `src/library` folder itself) to be more explicit about what kind of functionality they hold. To those of use working on the project, it's clear what these all refer to; however, to a future developer who is working on their first ticket in this repo, it might not be so obvious.

For example:
* `library` -> `libraries` or `library-authoring`
* `form` -> `create` or `create-library`
* etc.

arbrandes replied:
Good idea.  Done.

kdmccormick commented on src/library/common/data/constants.js:
```diff
@@ -0,0 +1,17 @@
+export const LIBRARY_TYPES = {
+  COMPLEX: 'complex', // blockstore/v2
+  LEGACY: 'legacy', // modulestore/v1
```
@marcotuts Are "complex" and "legacy" the user-facing names we want to move forward with to distinguish between V1 and V2 libraries? I believe the term "complex" came from when we were considering having separate video-specific or problem-specific library types. If that's no longer relevant, wouldn't it be clearer to just have, say, "Version 1 (Legacy)" and "Version 2" libraries?

arbrandes commented on .env:
```diff
@@ -5,6 +5,7 @@ CREDENTIALS_BASE_URL=null
 CSRF_TOKEN_API_PATH=null
 ECOMMERCE_BASE_URL=null
 LANGUAGE_PREFERENCE_COOKIE_NAME=null
+CMS_BASE_URL=null
```
No problem.  Done.

arbrandes commented on Makefile:
```diff
@@ -1,4 +1,4 @@
-transifex_resource = frontend-template-application
+transifex_resource = frontend-app-library-authoring
 transifex_langs = "ar,fr,es_419,zh_CN"
```
Probably a good idea, yes. :)

arbrandes commented on src/library/common/data/shapes.js:
```diff
@@ -0,0 +1,14 @@
+  description: PropTypes.string,
+  version: PropTypes.number,
+  has_unpublished_changes: PropTypes.bool,
+  has_unpublished_deletes: PropTypes.bool,
```
I've marked the ones that should always be available as required.

kdmccormick commented:
@arbrandes Great; let me know when you'd like the button pressed.

arbrandes commented:
@kdmccormick, I've addressed all the linting issues and renamed most of the other stuff according to your suggestions.  Unless you have any further comments, this should be good to merge.  Thanks!
